presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -230,15 +261,18 @@ public synchronized void freeRevocable(QueryId queryId, long bytes)
         checkArgument(queryReservation - bytes >= 0, "tried to free more revocable memory than is reserved by query");
         queryReservation -= bytes;
         if (queryReservation == 0) {
-            queryMemoryRevocableReservations.remove(queryId);
+            queryMemoryRevocableReservations.remove(queryId, queryReservation);
         }
         else {
             queryMemoryRevocableReservations.put(queryId, queryReservation);
```

Comment:
This still doesn't match how regular free works. This blind put overwrites the value another thread has put in. I think it needs to basically match what we did in free in terms of merging in the initial. Then you can remove the "else".

Reply:
I removed old `read and then remove`, and uses the strategy in `free`